utils/httputil/server.go

```diff
@@ -0,0 +1,50 @@
+package httputil
+
+import (
+	"context"
+	"net"
+	"net/http"
+	"time"
+)
+
+type grace interface {
+	Shutdown(context.Context) error
+}
+
+func GracefulListenAndServe(ctx context.Context, server *http.Server, shutdownTimeout ...time.Duration) error {
+	return gracefulFunc(ctx, server, server.ListenAndServe, shutdownTimeout...)
+}
+
+func GracefulServe(ctx context.Context, server *http.Server, l net.Listener, shutdownTimeout ...time.Duration) error {
+	fn := func() error {
+		return server.Serve(l)
+	}
+	return gracefulFunc(ctx, server, fn, shutdownTimeout...)
+}
```

Comment:
The reason I've introduced those two methods is: 
1. We are using graceful shutdown methods in many places
2. Hard and repetitive to test shutdown functionality when it is not in isolation.
3. It is easy to get the shutdown wrong; some of the implementations in our code were correct, and some were not (funny, I think I have introduced them all :D )

**Example of incorrect implementation:**
```go
	go func() {
		<-ctx.Done()
		_ = srv.Shutdown(context.Background())
	}()

	if err := srv.ListenAndServe(); err != nil {
		return fmt.Errorf("debug server: %w", err)
	}
	return nil
```
> When `Shutdown` is called `ListenAndServe` returns immediately, always with an error.

Two problems: 
1. The shutdown is the one that blocks and the one we should be waiting for.
2. An error is returned when the context gets cancelled.

semi-correct:
```go
	g, ctx := errgroup.WithContext(ctx)
	g.Go(func() error {
		<-ctx.Done()
		return srv.Shutdown(context.Background())
	})
	g.Go(func() error {
		return srv.ListenAndServe()
	})

	return g.Wait()
```
Here we wait for both go routines to return, but the `ListenAndServe` is still there.

The new implementation fixes both issues along with providing tests.

Reply:
Sounds good to me, although I'm not sure we need the `Graceful` part in the name. What do you think about just having it as `ListenAndServe` since we don't want an ungraceful one?

Reply:
Yeah, make sense :)